docs/rfc/007-endpoint-errors.md

```diff
@@ -0,0 +1,229 @@
+pros and cons, and picking one side or the other inevitably involves making sacrifices.
+
+**The case for mandatory handling**
+
+_Forcing_ consumers to handle your exceptional cases is beneficial to the _producer_. You can have confidence that
+consumers will be encouraged to handle the exceptional cases, even if they are not reading your documentation. Acknowledging
+the presence of exceptional cases will be required at the minimum.
+
+There are benefits as a consumer as well - additional errors, behaviors, or edge cases will be exposed to you on upgrading
+your client, and you should (in theory) never miss out on a new case (assuming you keep your client relatively up to date).
+
+**The case for optional handling**
+
+Optional handling favors the _consumer_ - existing code and behaviors will not have to change at the outset, and they
+can opt into improved error handling as necessary. The obvious downside is that producers could publish new errors until
+the cows come home, but lazy consumers may never actually handle these edge cases.
+
+## Implementation (Java)
+
+Declaring errors on endpoints not only allows producers to enforce that each error case is going to be handled by a 
+consumer, but it also allows us to provide the tools to make that ergonomic for consumers. Exactly how those ergonomics 
+will be provided depend on the language. There were two main approaches that were considered:
+
+**Enriched Types**
+
+With an expressive type system, it is possible to express the different possible outcomes as an enriched version of the 
+normal return type. This could be done, for example, by using union types, which already exist in the Conjure universe. 
+Some languages are more amiable to union types than others. 
+
+**Checked Exceptions**
+
+Java provides checked exceptions as a method of enforcing that exceptional states are handled, and declared errors on 
+endpoints fit nicely into that.
+
+**Why The Demo Uses Checked Exceptions**
+
+As of now, `SerializableError` does not lend itself to complex data types as complex error objects are flattened into a 
+`Map<String, String>`, which makes parsing it out very difficult. Checked exceptions preserve the existing status-quo 
+(that is, not messing with the reality of  `SerializableError`), but still improve expressivity and enforce errors being 
+actually handled. 
+
+## Open Questions ##
+
+1) Should handling errors on endpoints be mandatory? While it seems like a good idea in principle, could it cause 'bad'
+   behavior in client or server code?
+2) In the Java implementation, are checked exceptions the way to go? There are several other possible implementations
+   and ideas that may be more in-line with other languages, such as using a union type via a visitor pattern (which 
+   can be made much less verbose via a 'Visitor Builder').
+3) How important is implementation consistency across languages? Would it be acceptable to have union types in TS, but
```

Comment:
do we have a stance on this for conjure generally?

My inclination is to opt for whatever feels most natural in each language (e.g. I think in rust we model conjure unions as enums).

Reply:
I would vote for trying to keep things idiomatic in each potential target language.

Reply:
As a person who deals with many repos I personally prefer consistency over what feels natural in a language. The overhead of “unnatural” isn’t that high but overhead of having to change your mental model when you look at different repo is high in comparison from experience.